[62557][62607] date picker successor and children tabs must show the same correct chronological order

cbliard opened the pull request:
# Ticket

https://community.openproject.org/wp/62557
https://community.openproject.org/wp/62607

# What are you trying to accomplish?

The relations order was not the same between date picker tabs and relation tab for the same relation group.

Also the order of relation groups was not the same from one work package to another in the relations tab.

This needs to be consistent.

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

# What approach did you choose and why?

Move the ordering logic out of the view and put it in the `WorkPackageRelationsTab::RelationsMediator` object instead:
* `WorkPackageRelationsTab::RelationsMediator` returns all relation groups, child and relations, ordered as they should be
* `WorkPackageRelationsTab::RelationsMediator::RelationGroup` is ordering and returning `WorkPackageRelationsTab::RelationsMediator::RelationItem` objects as they should be
* `WorkPackageRelationsTab::RelationsMediator::RelationItem` have every information needed to do the rendering: visibility, relation, related work package, is the closest one, etc.

This is to alleviate most of the logic from the view, replacing conditionals with more specialized value objects.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

cbliard committed:
table_helpers: fix bug when defining multiple relations at once

ActiveSupport deep_merge does not concatenate arrays from different
hashes. Relations were stored as an array.

Switch to a hash to store all relations without any loss.

cbliard committed:
[62557] Fix relations order in date picker

https://community.openproject.org/wp/62557

Date picker tabs and relations tab are now sharing the same code for
ordering the relations so there should not be any difference about it
now.

myabc commented on app/components/work_package_relations_tab/relation_component.rb:
```diff
@@ -1,44 +1,69 @@
   # Checks if the relation or child work package is visible to the current user
   #
   # @param work_package [WorkPackage] The work package whose relations are being displayed
   # @param relation [Relation, nil] The relation between work packages, if any
```
Can you document the `@param relation_item` here too?

cbliard replied:
Will do.
It's still a draft at the moment. I just want the test suite to run.

cbliard committed:
refactor: use RelationItem attributes directly

Instead of using parameters given to the RelationComponent.

cbliard committed:
Fix flaky spec

Failing spec is `./modules/meeting/spec/features/structured_meetings/structured_meeting_crud_spec.rb:90`
in job https://github.com/opf/openproject/actions/runs/14190043033/job/39752454726

When reordering agenda items, the order is not updated immediately under
high load. If the action menu is opened before the reordering is
finished, the action menu will close when the items are reordered and
clicking on the item "Agenda item" will fail.

The flaky spec is fixed by waiting for the order to be updated with
`assert_agenda_order!`.

cbliard committed:
refactor: use information from RelationGroup directly for rendering

Also:
- Precompute the closest relation in the `RelationGroup` constructor to
  avoid recomputing it every time.
- Move the `closest?` information in the `RelationItem` instance
  directly to remove its computation from the component.html logic.

cbliard committed:
[62607] Always display relations in the relations tab in a fixed order

https://community.openproject.org/wp/62607

The order of the relation groups was not consistent from one work
package to another. It's because the relations are returned in any order
from the database.

Instead of getting relations from database and then grouping them by
type as they come up in the array, get the types in the order we want
and build up the relation groups from there.

There are no differences between rendering a `RelationGroup` of type
child and a `RelationGroup` of type relation anymore. For this to
happen, the "children" type had to be renamed "child", which is in line
with the Relation::TYPE_CHILD constant.

cbliard committed:
table_helpers: Add successor relations column type

Example usage:
  | subject      | successors                   |
  | predecessor2 | main, predecessor with lag 2 |
  | predecessor  | precedes main                |
  | main         |                              |

myabc commented on app/components/work_package_relations_tab/relations_mediator.rb:
```diff
@@ -46,16 +58,59 @@ def all_relations
+  # @param work_package [WorkPackage] The work package for which the relation
+  #   is displayed
+  # @param related [WorkPackage] The related work package: the other work
+  #   package of therelation, or the child for child relations.
```
missing space: `therelation`

myabc commented on app/models/relation.rb:
```diff
@@ -49,6 +49,9 @@ class Relation < ApplicationRecord
   TYPE_PARENT       = "parent"
   TYPE_CHILD        = "child"
 
+  # The order of the types is important. It's used to build up `ORDERED_TYPES`
+  # which is used to order relations of different kind like in the "Add relation"
+  # menu or the relations tab.
```
👍🏻

myabc commented on app/components/work_package_relations_tab/relation_component.rb:
```diff
@@ -1,44 +1,65 @@
-  attr_reader :work_package, :relation, :visibility, :child, :editable
+  attr_reader :relation_item, :editable
+
+  delegate :relation, :work_package, to: :relation_item
```
```suggestion
  delegate :relation, :work_package, :visible?, :closest?, to: :relation_item
```

cbliard replied:
Adopted.

I have a love-hate relation with `delegate`. I find it make the code harder to scan, and may also hide some unwanted chained calls (for instance `relation.description.present?` is actually `relation_item.relation.description.present?`) but on the other hand, it's simply convenient and conveys the idea that some methods are delegated.

Not sure why I have that feeling though.

myabc commented on app/components/work_package_relations_tab/relation_component.rb:
```diff
@@ -1,44 +1,65 @@
+  def child
+    if parent_child_relationship?
+      relation_item.related
+    end
```
```suggestion
   relation_item.related if parent_child_relationship?
```

reads a bit better to me.

cbliard replied:
That's funny. Actually it does not read better for me.

When I scan it, I tend to omit the `if` and I think that it always returns a value. It's only once I read it that I see the `if`. That's why most of the time I reserve one-line if for guard clauses.

What would read better for me would be something like this:

```
return unless parent_child_relationship?

relation_item.related
```

Also, I could inline the `child` and replace its call with `relation_item.related` directly but I found `child` more intention-revealing.

What do you think?

cbliard replied:
I used the code you suggested.

myabc commented on app/components/work_package_relations_tab/relation_component.rb:
```diff
@@ -1,44 +1,65 @@
   include OpPrimer::ComponentHelpers
 
-  attr_reader :work_package, :relation, :visibility, :child, :editable
+  attr_reader :relation_item, :editable
```
While working on this part of the codebase a few weeks ago, I hesitated about getting rid of the `editable` accessor. For the sake of consistency, I would prefer to define an `editable?` predicate method. That said, `editable` appears to only be called once - from `should_render_action_menu?` - so we could forgo the method completely.

cbliard replied:
I prefer `editable?` too. I added the method and kept the `attr_reader`.

myabc commented on app/components/work_package_relations_tab/relations_mediator.rb:
```diff
@@ -46,16 +58,59 @@ def all_relations
 
-    def closest_relation
-      return nil unless type.follows?
+    def to_relation_items(relations, visibility)
```
this method could be made private.

cbliard replied:
Actually most of the class could be made private. The methods are used only to build the relation items.

I have an experiment on another branch that I just pushed where I introduced a `RelationGroupBuilder` where all these methods are grouped, and it builds up a `RelationGroup` which only has `type` and `relation_items` left.

Could you review it and tell me if I'm pushing it too far or if it's worth adding it? The commit is here: https://github.com/opf/openproject/commit/b5b64f7cce69ec93393fe7005c423cb5c60e3491

myabc commented on app/components/work_package_relations_tab/relations_mediator.rb:
```diff
@@ -29,9 +29,21 @@
+  RelationGroup = Data.define(:type, :work_package, :visible_relations, :ghost_relations, :closest_relation) do
+    def initialize(type:, work_package:, visible_relations:, ghost_relations:)
+      type = ActiveSupport::StringInquirer.new(type.to_s)
+      closest_relation = self.class.closest_relation(type, visible_relations + ghost_relations)
```
👍🏻 good that we're only calculating the closest relation once on instantiation (that was my mistake).

We could remove the type param from `self.closest_relation` and check the type here instead.

```suggestion
      closest_relation = self.class.closest_relation(visible_relations + ghost_relations) if type.follows?
```

myabc commented on app/components/work_package_relations_tab/relations_mediator.rb:
```diff
@@ -29,9 +29,21 @@
+      super(type:, work_package:, visible_relations:, ghost_relations:, closest_relation:)
+    end
+
+    def self.closest_relation(type, relations)
+      return nil unless type.follows?
+
```
```suggestion
   def self.closest_relation(relations)
```

We could also make this a `private_class_method`.

myabc replied:
That said, it might be even more sense to move the closest logic to `ClosestRelation`, e.g.

```ruby
    def self.of(relations)
      relations
        .map { new(it) }
        .select(&:soonest_start)
        .max
        &.relation
    end
```

cbliard replied:
I like the idea of moving it to `ClosestRelation`. I'll try that.

myabc commented on spec/components/work_package_relations_tab/relation_component_spec.rb:
```diff
@@ -43,92 +43,119 @@
+
+      context "when not editable" do
+        it "does not render an action menu" do
+          component = render_component(relation_item:, editable: false)
+          expect(component).not_to have_menu
```
👍🏻

myabc commented on spec/components/work_package_relations_tab/relations_mediator_spec.rb:
```diff
@@ -94,5 +129,33 @@
+      it "returns all relations of the group as RelationItem instances, " \
+         "ordered by oldest first (lowest id first), mixing visible and ghost relations" do
+        relation_group = described_class::RelationGroup.new(
+          type: "relates",
```
would it be more consistent to use the constant here?

```suggestion
          type: Relation::TYPE_RELATES,
```

cbliard replied:
yes, definitely.
Actually, I changed "children" to "child", then I looked into `Relation` to see how to implement the order and saw that constant `TYPE_CHILD`, so I thought I should use it, and then the editor suggested to replace more of them which seemed like a good idea, but I did not replace them all.

That's good that you spotted this one.

myabc reviewed (commented):
I haven't had a chance to look at the table helper code just yet, but thought I would go ahead and give feedback on the substantive part of the PR:

Overall this is a significant improvement to `RelationsComponent`, `RelationsMediator` and associated classes. It makes the code a lot easier to reason about.

I have to admit that I don't particularly like `RelationItem` as a name (it's generic and not immediately clear how it differs from `Relation`). Having said that, I don't have a better suggestion for the moment. The fact that it is a nested class within the `WorkPackageRelationsTab` module does make it clear that it is UI code.

myabc commented:
Ideally there would also be some basic component spec coverage for the `DialogContentComponent` relations tabs. `dialog_content_component_spec.rb` really only tests `DatePicker::FormComponent`.

cbliard committed:
refactor: Integrate PR review comments

Most notable is that computation of closest relation has been moved to
the ClosestRelation class.

Co-authored-by: Alexander Brandon Coles <a.coles@openproject.com>
